db/compaction_iterator.cc

```diff
@@ -623,4 +630,12 @@ inline SequenceNumber CompactionIterator::findEarliestVisibleSnapshot(
   return kMaxSequenceNumber;
 }
 
+// used in 2 places - prevents deletion markers to be dropped if they may be
+// needed and disables seqnum zero-out in PrepareOutput for recent keys.
+inline bool CompactionIterator::ikeyNotNeededForIncrementalSnapshot() {
+  return (!compaction_->preserve_deletes()) ||
+         (preserve_deletes_seqnum_ == nullptr) ||
+         (ikey_.sequence < preserve_deletes_seqnum_->load());
```

Comment:
Not sure I followed.

Assuming the preserve_deletes comprise monotonically increasing sequence, let's say that `preserve_deletes_seqnum_` was s1, and ikey `k1` had seqnum s2 < s1 (so we decided to skip it). Then if the preserved_seqnum was increased, the invariant s2 < s1 still holds true right? I.e. if before preserve_seqnum was increased the key was eligyable for skipping (so we drop it) it would still be fine to drop if we bump the seqnum?

Reply:
Sorry my explanation above was unclear -- I meant skip dropping a tombstone, so actually preserving it :p. Following your example, the problem would be when:

- s2 >= s1 so we preserve a tombstone
- Then user increases s1 such that s1 > s2
- Now `PrepareOutput` decides to zero the tombstone's seqnum

Reply:
Oh I see. Hm yeah. Thinking about options...

Why do we need this assert here?

Reply:
Well, previously writing out a tombstone with seqnum zero was considered a bug because it couldn't possibly cover any data. This assertion has been fairly useful as it caught a bug for me within the past two weeks.

One way you can fix is store a local copy of the current `preserve_deletes_seqnum_` in a CompactionIterator instance variable when you decide whether to drop a delete. Then use that variable when you decide whether to zero its seqnum.

Reply:
You could consider simplifying further and make the whole `CompactionIterator` use a constant value for `preserve_deletes_seqnum_` that is retrieved when the iterator is constructed.